WIP: Search in Card Browser for highlighted text from Review mode

Profpatsch opened the pull request:
I tried my hands at https://github.com/ankidroid/Anki-Android/issues/4496, but that’s as far as I can get. See commit message for the problem.
Basically I can’t find a way to add the callbacks to the action mode. Also, I have no idea what action mode I’m looking at.

I thought about tracking for long presses, but then the question becomes how to still get the text selection functionality.

@timrae: Maybe you have an idea how to implement this functionality?

timrae commented:
@Profpatsch 

Thanks for having a go at this. It took me a while to find the article, but [this is the way](https://medium.com/google-developers/custom-text-selection-actions-with-action-process-text-191f792d2999) that I was imagining it should be implemented in my comment on #4496. By default that would add an item to the floating action bar whenever text is selected inside **any app**, which may not be what users want. But according to the comments section of that article, an [issue](https://issuetracker.google.com/issues/37078396) was added and subsequently fixed that would allow you to limit the visibility of the intents to your own app if you set exported to false, which I think we should do.

timrae commented:
FYI, [this](https://stackoverflow.com/questions/30518089/how-to-use-the-new-android-m-feature-of-text-selection-to-be-offered-from-outs/30518324) may also be helpful, though I'm not sure how up-to-date it is.

Profpatsch commented:
Okay, I rewrote it to a very simple implementation based on `ACTION_PROCESS_TEXT`.

It’s still visible from other apps because the hiding functionality is not in Android yet. I also don’t know if I have implemented it in a backwards-compatible manner, since I only have a new device to test it on.

timrae commented:
Normally we just test with the simulator during development

timrae commented:
`setHandleNativeActionModesEnabled` is available on 4.0 and above. I think we're finally ready to bump the minimum supported version up to 4.1 for v2.9, but we'd better announce it on the forums.

Profpatsch commented:
Okay, I moved the `ACTION_PROCESS_TEXT`-specific logic into a `Compat` method.

timrae commented:
Almost, but that's not quite going to work; you need to instantiate the correct Compat class for the correct API using CompatHelper:

https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.java

Profpatsch commented:
Right. Added v23 to the dispatcher and switched to the singleton method .

Profpatsch commented:
@timrae Any further remarks?

timrae commented:
Thanks for reminding me. Have you tested it properly with the simulator
running Android 4.1? I'm also still unsure about this `exported` issue...
When was support for this added into the Android framework? One of us will
need to check the framework code.

I need to make a thread on the forum about the change in minimum API, and
we should probably make a maintenance release soon before dropping support
for 2.3.

On 4/10/2017 01:47, "Profpatsch" <notifications@github.com> wrote:

> @timrae <https://github.com/timrae> Any further remarks?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ankidroid/Anki-Android/pull/4708#issuecomment-333907042>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACsA4ubBc3sRoiwFvucpcWC4WwZha4Pzks5somU6gaJpZM4PdL3B>
> .
>

timrae commented:
I've made the [announcement](https://groups.google.com/forum/#!topic/anki-android/jyK9LJG8rpY) that *Android 4.0* will be the new minimum going forward. We can bump the minimum again to 4.1 whenever the need arises.

Profpatsch commented:
> I'm also still unsure about this `exported` issue...
> When was support for this added into the Android framework? One of us will
> need to check the framework code.

I’m actually not sure if it wouldn’t be better to export it. That way, if you find e.g. a word in your browser, you can highlight it and directly see if you already have it in your vocab list, which is awesome.
The item text uses the `Activity` name, so it might be sensible to rename it to “Vocab Browser” for example?

timrae commented:
> I’m actually not sure if it wouldn’t be better to export it. That way, if you find e.g. a word in your browser, you can highlight it and directly see if you already have it in your vocab list, which is awesome.

This sounds like it could be useful if we could streamline the process of note addition from the browser a bit better, so that when you choose "Add note" from the browser while there is something in the search box, it automatically populates that text into the first field of the note editor.

timrae commented:
> The item text uses the Activity name, so it might be sensible to rename it to “Vocab Browser” for example?

There's no way to customize it without changing the Activity name?

> Update: Right now the browser crashes once I click “Card Browser” though, probably a SecurityException?

What's the stacktrace?

Profpatsch commented:
> There's no way to customize it without changing the Activity name?

At least that’s what the [article](https://medium.com/google-developers/custom-text-selection-actions-with-action-process-text-191f792d2999) says:
![screenshot](https://user-images.githubusercontent.com/3153638/31155550-66afeea0-a8af-11e7-9a10-a46d8041d439.png)

I can’t find any documentation other than https://developer.android.com/reference/android/content/Intent.html#ACTION_PROCESS_TEXT and that medium article; apparently Medium is the way the Android team distributes their docs?


> What's the stacktrace?

```
10-04 02:57:31.336  2182 22848 I ActivityManager: START u0 {act=android.intent.action.PROCESS_TEXT typ=text/plain cmp=com.ichi2.anki/.CardBrowser (has extras)} from uid 10050 on display 0
10-04 02:57:31.336  2182 22848 W ActivityManager: Permission Denial: starting Intent { act=android.intent.action.PROCESS_TEXT typ=text/plain cmp=com.ichi2.anki/.CardBrowser (has extras) } from ProcessRecord{5ad81a9 26760:org.lineageos.jelly/u0a50} (pid=26760, uid=10050) not exported from uid 10069
10-04 02:57:31.339 26760 26760 D AndroidRuntime: Shutting down VM
10-04 02:57:31.340 26760 26760 E AndroidRuntime: FATAL EXCEPTION: main
10-04 02:57:31.340 26760 26760 E AndroidRuntime: Process: org.lineageos.jelly, PID: 26760
10-04 02:57:31.340 26760 26760 E AndroidRuntime: java.lang.SecurityException: Permission Denial: starting Intent { act=android.intent.action.PROCESS_TEXT typ=text/plain cmp=com.ichi2.anki/.CardBrowser (has extras) } from ProcessRecord{5ad81a9 26760:org.lineageos.jelly/u0a50} (pid=26760, uid=10050) not exported from uid 10069
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:1684)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:1637)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.app.ActivityManagerProxy.startActivity(ActivityManagerNative.java:3101)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.app.Instrumentation.execStartActivity(Instrumentation.java:1641)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.app.Activity.startActivityForResult(Activity.java:4879)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.content.ContextWrapper.startActivityForResult(ContextWrapper.java:368)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.view.View.startActivityForResult(View.java:5845)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.widget.Editor$ProcessTextIntentActionsHandler.fireIntent(Editor.java:6246)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.widget.Editor$ProcessTextIntentActionsHandler.performMenuItemAction(Editor.java:6200)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.widget.Editor$TextActionModeCallback.onActionItemClicked(Editor.java:3786)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at com.android.internal.policy.DecorView$ActionModeCallback2Wrapper.onActionItemClicked(DecorView.java:2303)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at com.android.internal.view.FloatingActionMode$3.onMenuItemSelected(FloatingActionMode.java:88)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at com.android.internal.view.menu.MenuBuilder.dispatchMenuItemSelected(MenuBuilder.java:761)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at com.android.internal.view.menu.MenuItemImpl.invoke(MenuItemImpl.java:152)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at com.android.internal.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:904)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at com.android.internal.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:894)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at com.android.internal.view.FloatingActionMode$4.onMenuItemClick(FloatingActionMode.java:114)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at com.android.internal.widget.FloatingToolbar$FloatingToolbarPopup$15.onItemClick(FloatingToolbar.java:1379)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.widget.AdapterView.performItemClick(AdapterView.java:310)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.widget.AbsListView.performItemClick(AbsListView.java:1164)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.widget.AbsListView$PerformClick.run(AbsListView.java:3139)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.widget.AbsListView$3.run(AbsListView.java:4054)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:751)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:95)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:154)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6186)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
10-04 02:57:31.340 26760 26760 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
10-04 02:57:31.343  2182 22597 W ActivityManager:   Force finishing activity org.lineageos.jelly/.MainActivity
```

timrae commented:
Hmmm, this stack trace was generated with `exported=true`?

> Permission Denial: starting Intent { act=android.intent.action.PROCESS_TEXT typ=text/plain cmp=com.ichi2.anki/.CardBrowser (has extras) } from ProcessRecord{5ad81a9 26760:org.lineageos.jelly/u0a50} (pid=26760, uid=10050) not exported from uid 10069

timrae commented:
I'm a bit confused, as I would have thought that on older versions of Android the `PROCESS_TEXT` intent is simply never sent...

Profpatsch commented:
> Hmmm, this stack trace was generated with exported=true?

I understand now, I should have changed the `exported` attribute to `true`. Once I did, it works.

timrae commented:
Instead of that wrapper function inside the Compat classes, I think it might be more simple to just port over the missing constants [from the framework](https://github.com/android/platform_frameworks_base/blob/nougat-release/core/java/android/content/Intent.java#L3208).

```
public class CompatV10 implements Compat {
    public static final String ACTION_PROCESS_TEXT = "android.intent.action.PROCESS_TEXT";
    public static final String EXTRA_PROCESS_TEXT = "android.intent.extra.PROCESS_TEXT";
    // ....
}

public class CompatV23 implements Compat {
    public static final String ACTION_PROCESS_TEXT = Intent.ACTION_PROCESS_TEXT;
    public static final String EXTRA_PROCESS_TEXT = Intent.EXTRA_PROCESS_TEXT;
}
```

timrae commented:
> I understand now, I should have changed the exported attribute to true. Once I did, it works.

Ah right, yes, the activity itself...

Profpatsch commented:
> Instead of that wrapper function inside the Compat classes, I think it might be more simple to just port over the missing constants from the framework.

> Ah right, yes, the activity itself...

So no need to copy the strings, right?

timrae commented:
> So no need to copy the strings, right?

Sorry I'm not following your question... Did you see the code sample I posted above?

Profpatsch commented:
> Sorry I'm not following your question... Did you see the code sample I posted above?

Yeah, but I’m not quite sure what this would simplify and how. Do you mean I should remove [this function](https://github.com/ankidroid/Anki-Android/pull/4708/files#diff-9bd267b573e99a15cac3d534a7b8ab87R13) alltogether?

timrae commented:
Yes, the only part of the function that is not available on all APIs is the two constants, right?

Profpatsch commented:
Ah! Now I get it. Should have gone to bed earlier yesterday. :)
Will do.

Profpatsch commented:
I implemented it; the final strings have to be in the interface, otherwise they cannot be used with `getCompat()` (which returns the interface). I refuse to cast and break the abstraction. ;)

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:
```diff
@@ -622,10 +623,15 @@ public void onClick(View v) {
+        Compat compat = CompatHelper.getCompat();
+        if (intent.getAction() == compat.ACTION_PROCESS_TEXT) {
+            CharSequence search = intent.getCharSequenceExtra(compat.EXTRA_PROCESS_TEXT);
+            assert search != null;
```
Thanks, this is much clearer! I think we don't really want to crash if an app developer accidentally doesn't add `EXTRA_PROCESS_TEXT`... Can you get rid of the assert and just write the following?

`if (search != null && search.length() != 0) {`

Last nitpick I promise :)

Profpatsch replied:
> I think we don't really want to crash if an app developer accidentally doesn't add EXTRA_PROCESS_TEXT

That’s something that should never happen, which is why I use an assert.
From the `ACTION_PROCESS_TEXT` documentation:

```
Activity Action: Process a piece of text.
Input: EXTRA_PROCESS_TEXT contains the text to be processed. EXTRA_PROCESS_TEXT_READONLY states if the resulting text will be read-only.
Output: EXTRA_PROCESS_TEXT contains the processed text.
```

timrae replied:
> From the ACTION_PROCESS_TEXT documentation

It doesn't actually say anywhere there that it won't be null, or that it's guaranteed to be present. In fact it will only be there if the developer sending the intent calls `putExtra()`. You can see this from the `getCharSequenceExtra()` documentation, which says:

> (returns) the value of an item that previously added with putExtra() or null if no CharSequence value was found.

timrae replied:
FYI: In general, we try to only use `assert()` to pick up programming errors *by us*, whereas this would be picking up programming errors by others (i.e. other apps, android framework).

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:
```diff
@@ -62,6 +62,9 @@
 import com.ichi2.async.DeckTask.TaskData;
+import com.ichi2.compat.Compat;
+import com.ichi2.compat.CompatHelper;
+import com.ichi2.compat.CompatV23;
```
Oh I guess this import isn't needed anymore btw

timrae commented on AnkiDroid/src/main/java/com/ichi2/compat/CompatV23.java:
```diff
@@ -0,0 +1,15 @@
+import android.annotation.TargetApi;
+import android.content.Intent;
+
+import timber.log.Timber;
```
Also not needed

Profpatsch replied:
The compiler doesn’t show these as warnings … :(

timrae replied:
Yeah, most compilers don't unfortunately, even in other programming languages...You can run "optimize imports" with Android Studio which will take care of this for you automatically.

Profpatsch commented:
Unneeded imports & `assert` have been fixed.

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:
```diff
@@ -617,6 +619,19 @@ public void onClick(View v) {
+        if (intent.getAction() == compat.ACTION_PROCESS_TEXT) {
+            CharSequence search = intent.getCharSequenceExtra(compat.EXTRA_PROCESS_TEXT);
+            if (search != null && search.length() != 0) {
+                Timber.d("CardBrowser :: Called with search intent", search);
```
I guess you forgot to add the `%s` formatting string here so that the content of `search` gets printed out...?

Profpatsch replied:
Oh, it’s a format string. Wasn’t clear from the documentation, thought it would display them automatically in a smart way.

timrae commented:
Are you interested in implementing the idea I posted above to allow adding a new note from the content of the browser search bar?

Profpatsch commented:
> Are you interested in implementing the idea I posted above to allow adding a new note from the content of the browser search bar?

At the moment I’m short on time, but I’ll put it on my side-project list.
